pnet_sys/src/windows.rs

```diff
@@ -1,112 +1,224 @@
+                Ok(SocketAddr::V4(sockaddrv4))
+            }
+            AF_INET6 => {
+                assert!(len as usize >= mem::size_of::<SockAddrIn6>());
+                let storage: &SockAddrIn6 = mem::transmute(storage);
+                let arr: [u16; 8] = mem::transmute(*storage.sin6_addr.u.Word());
+                let a = ntohs(arr[0]);
+                let b = ntohs(arr[1]);
+                let c = ntohs(arr[2]);
+                let d = ntohs(arr[3]);
+                let e = ntohs(arr[4]);
+                let f = ntohs(arr[5]);
+                let g = ntohs(arr[6]);
+                let h = ntohs(arr[7]);
+                let ip = Ipv6Addr::new(a, b, c, d, e, f, g, h);
+                Ok(SocketAddr::V6(SocketAddrV6::new(ip,
+                                                    ntohs(storage.sin6_port),
+                                                    u32::from_be(storage.sin6_flowinfo),
+                                                    *storage.u.sin6_scope_id())))
+            }
+            _ => Err(io::Error::new(io::ErrorKind::InvalidData, "expected IPv4 or IPv6 socket")),
+        }
+        }
+        
     }
 }
 
 use self::public::*;
 
 #[inline(always)]
 pub fn ipv4_addr(addr: InAddr) -> u32 {
-    (addr.S_un as u32).to_be()
+    unsafe {
+        (*addr.S_un.S_addr() as u32).to_be()
+    }
 }
 
 #[inline(always)]
 pub fn mk_inaddr(addr: u32) -> InAddr {
-    InAddr { S_un: addr as winapi::ULONG }
+    let mut val = InAddr::default();
+    unsafe {
+        *val.S_un.S_addr_mut() = addr as minwindef::ULONG;
+    }
+    val
 }
 
-
 pub unsafe fn sendto(socket: CSocket,
                      buf: Buf,
                      len: BufLen,
-                     flags: libc::c_int,
+                     flags: ctypes::c_int,
                      to: *const SockAddr,
                      tolen: SockLen)
     -> CouldFail {
-    ws2_32::sendto(socket, buf, len, flags, to, tolen)
+    winsock2::sendto(socket, buf, len, flags, to, tolen)
 }
 
+
 pub unsafe fn recvfrom(socket: CSocket,
                        buf: MutBuf,
                        len: BufLen,
-                       flags: libc::c_int,
+                       flags: ctypes::c_int,
                        addr: *mut SockAddr,
                        addrlen: *mut SockLen)
     -> CouldFail {
-    ws2_32::recvfrom(socket, buf, len, flags, addr, addrlen)
+    winsock2::recvfrom(socket, buf, len, flags, addr, addrlen)
 }
 
+pub fn make_in6_addr(segments: [u16; 8]) -> In6Addr {
+    let mut val = In6Addr::default();
+    unsafe {
+        *val.u.Word_mut() = [htons(segments[0]),
+                        htons(segments[1]),
+                        htons(segments[2]),
+                        htons(segments[3]),
+                        htons(segments[4]),
+                        htons(segments[5]),
+                        htons(segments[6]),
+                        htons(segments[7])];
+    }
+    
+    // val = unsafe {
```

Comment:
Commented code can be removed :)

Reply:
I left it in there, and forgot to comment it with a question. The original code uses mem::transmute(). I'm pretty new to Rust, so I won't presume I know better than others. It made me wonder, based on rust docs and guides, if mem::transmute can be removed entirely from libpnet, same with mem::uninitialized. I'd like to know more about the use cases for that, when its preferred and why. My code replacement is still unsafe, due to the Word_mut() method being unsafe, but it seems safer than transmute, right?